---
title: Week4-day5 :integration template"
author: "Devika Agarwal/updataed by Carla Cohen"
date: "17/10/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) #print code and results
```


```{r}
library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(patchwork)
library(tidyverse)
library(batchelor)
library(harmony)
library(lisi)
library(MAST)
options(future.globals.maxSize = 4000 * 1024^2)    # Maximum allowed total size (in bytes) of global variables identified
```

# Exercise 1 - joint analysis of two PBMC datasets

Here we will look at how to PBMC datasets compare without any integration. 

## 1. Import the two RDS files into R

These RDS object have already been put through the QC and filtering steps covered on Seurat day 1.


```{r}
pbmcv2 <- readRDS("/ceph/project/obds/shared/resources/4_r_single_cell/singlecell_integration/pbmcv2_filtered.rds")
pbmcv3 <- readRDS("/ceph/project/obds/shared/resources/4_r_single_cell/singlecell_integration/pbmcv3_filtered.rds")
```


Plot these metadata covariates nCount_RNA/nFeature_RNA/percent.mt to see how they range for both the datasets 

use `VlnPlot()` from Seurat

```{r}
head(pbmcv2[[]])
p1 <- VlnPlot(pbmcv2, features=c("nCount_RNA", "nFeature_RNA", "percent.mt"))
p2 <- VlnPlot(pbmcv3, features=c("nCount_RNA", "nFeature_RNA", "percent.mt"))
cowplot::plot_grid(p1, p2, ncol=1)
```



## 2. Create a merged Seurat object containing both sets of data (no batch correction)

Vignette: https://satijalab.org/seurat/articles/merge_vignette.html

```{r}
seurat_merged <- merge(x=pbmcv2,y=pbmcv3, add.cell.ids=c("v2", "v3"))

```

Check the dims of the merged object in comparison to the separate objects

What is the total gene number in the merged object compared to each single object?

```{r}
dim(seurat_merged)
dim(pbmcv2)
dim(pbmcv3)
seurat_merged[[]]
```

## 3.1 Perform SCTransform normalisation

Do you want to regress any meta.data variables ?

```{r, warning=FALSE}
seurat_merged <- SCTransform(seurat_merged, assay = "RNA", new.assay.name = "SCT", vars.to.regress = "percent.mt")

```

Plot to see how the nCount_RNA looks before and after SCT normalisation for the merged dataset?

Did SCT do a good job in normalizing the sequencing depth across the two datasets?


```{r}
p3 <- VlnPlot(seurat_merged, features=c("nCount_RNA", "nCount_SCT"), same.y.lims = T)
p3
```


What should the Default Assay be?

```{r}
DefaultAssay(seurat_merged)
```

Plot the Variable features plot for data with top 10 variable genes labelled

```{r, fig.width=8}
plot_vf <- VariableFeaturePlot(seurat_merged, cols = c("black", "red"), assay = "SCT", selection.method = "sct")

vf <- VariableFeatures(seurat_merged)[1:10]
LabelPoints(plot_vf, points = vf, repel = T)
```

## 3.2 Run PCA and visualize 

Can you see a batch effect between pbmc v2 and v3?


```{r}
seurat_merged <- RunPCA(seurat_merged, assay="SCT")
```

Look at which genes are driving your PCs.

```{r, fig.width=8}
VizDimLoadings(seurat_merged, nfeatures = 20, balanced = T) #the further the zero the more influence that gene is to the PC; balanced = T to display both positive and negative influencing genes
```


```{r}
Reductions(seurat_merged)
pca_merge <- DimPlot(seurat_merged, dims = c(1,2), reduction = "pca")
pca_merge
```

Choose optimal  number of PCs for UMAP

```{r}
ElbowPlot(seurat_merged, ndims = 50)
```

## 3.3 Run the Seurat UMAP function and visualize

Is there a batch effect? 


```{r}
seurat_merged <- RunUMAP(seurat_merged, dims = 1:20)
umap_merge <- DimPlot(seurat_merged, dims = c(1,2))
umap_merge
```

### Optional: Do cells cluster by celltypes?

Plot some pbmc celltype markers to see how they look on the UMAP

```{r}
FeaturePlot(seurat_merged, features = c("CD3D", "CD3E", "GNLY", "LYZ", "S100A8", "S100A9", "CD79A", "CD79B", "MS4A1"), reduction = "umap")
```


## Exercise 2: Seurat integration with SCTransform

Check out the Vignette: https://satijalab.org/seurat/articles/integration_introduction.html#performing-integration-on-datasets-normalized-with-sctransform-1 

The Seurat v2 integration method is slightly different depending on whether you perform log normalistaion or SCTransformation on your seurat objects. Here we are using the SCTransform method (the vignette covers both).

### 1. Set up

Import the RDS files 

We will read in the objects again to start the integration workflow from scratch

```{r}
pbmcv2 <- readRDS("/ceph/project/obds/shared/resources/4_r_single_cell/singlecell_integration/pbmcv2_filtered.rds")

pbmcv3 <- readRDS("/ceph/project/obds/shared/resources/4_r_single_cell/singlecell_integration/pbmcv3_filtered.rds")
```

Generate a named list of the two objects to run workflow

```{r}
pbmc.list <- list("v2" = pbmcv2, "v3" = pbmcv3)
pbmc.list
```

Apply SCTransform across the list

```{r}
pbmc.list <- lapply(pbmc.list, SCTransform, assay = "RNA", new.assay.name = "SCT", vars.to.regress = "percent.mt") 
#previously with the merged dataset we performed SCT on all the cells at once, but with integration we do that on two separate data sets. 

```

Check dims for both objects in list after SCTransform
```{r}
sapply(pbmc.list, dim)
```

Make sure Default assay in both objects is SCT
```{r}
sapply(pbmc.list, DefaultAssay)

#if it wasn't sct, assing by:
#DefaultAssay(pbmc.list[[1]]) <- "SCT"
```

Choose the features to use when integrating multiple datasets. 

This function ranks features by the number of datasets they are deemed variable in, breaking ties by the median variable feature rank across datasets. It returns the top scoring features by this ranking.
```{r}
features <- SelectIntegrationFeatures(pbmc.list, nfeatures = 2000, assay=c("SCT", "SCT")) #looking for features that are variable across the two datasets
```

Prepare an object list normalized with sctransform for integration (run `PrepSCTIntegration()` )

This function subsets the scale.data slot to only contain residuals for anchor.features for efficiency in downstream processing
It is specific to the SCTransform workflow and makes sure you have the SCTransform residuals for the remaining genes

```{r}
pbmc.list <- PrepSCTIntegration(pbmc.list, assay = c("SCT", "SCT"), anchor.features = features)

```

> After this step, the features present in the scale.data slots of the two objects are the same
> anchor.features are used in the anchor finding process


### Optional exercise

How similar are these features to the variable features that we found for the pbmcv2 and v3 individually?
- extract variable features for each dataset individually
- compare these with "features"; which has the variable features most common for both datasets


## 2. Perform integration

Find a set of anchors between a list of Seurat objects. These anchors can later be used to integrate the objects.

```{r}
pbmc.anchors <- FindIntegrationAnchors(pbmc.list, assay = c("SCT", "SCT"), normalization.method = "SCT", anchor.features = features, dims = 1:20)

```

Integrate the two pbmc objects

```{r}
pbmc.combined.sct <- IntegrateData(anchorset = pbmc.anchors, new.assay.name = "integrated", normalization.method = "SCT", dims = 1:20)
```

What is the Default assay now?
Check to see if the Seurat object has changed in the Assay slot at all

```{r}
DefaultAssay(pbmc.combined.sct)
pbmc.combined.sct[[]] 
```

## 3. Analysis of integrated data

Run  PCA on Integrated data

```{r}
pbmc.combined.sct <- RunPCA(pbmc.combined.sct)
```

How many PCs should we use going forward?
```{r}
ElbowPlot(pbmc.combined.sct, ndims = 50)
```

Visualise PCA, how does this compare to the merged data?

```{r, fig.width=5, fig.height=2}
pca_int <- DimPlot(pbmc.combined.sct, dims = c(1,2), reduction = "pca")
cowplot::plot_grid(pca_merge, pca_int, ncol=2, labels = c("Merged", "Integrated"))
```



Run UMAP 
```{r}
pbmc.combined.sct <- RunUMAP(pbmc.combined.sct, reduction = "pca", dims = 1:20)

```


```{r}
DefaultAssay(pbmc.combined.sct)
```

Find neighbors and clusters for the integrated data

Run `FindNeighbors()` and `FindClusters()` 
Try using multiple resolutions at the same time for the clustering e.g. 0.5 and 0.8
What other clustering algorithms could be used here? Which one is used by default?


```{r}
pbmc.combined.sct <- FindNeighbors(pbmc.combined.sct, reduction = "pca", dims = 1:20)

pbmc.combined.sct <- FindClusters(pbmc.combined.sct, resolution = c(0.5, 0.8))
```


How many clusters have been defined at each resolution?
```{r}
levels(pbmc.combined.sct$integrated_snn_res.0.5) #9 clusters
levels(pbmc.combined.sct$integrated_snn_res.0.8) #11 clusters

#alternatively
unique(pbmc.combined.sct[[]]$integrated_snn_res.0.5)

# returns the number of cells in each cluster
table(pbmc.combined.sct$integrated_snn_res.0.5)

```

Visualse UMAP plot for batch effect and clusters

check the Default Assay to be sure

```{r, fig.width=5, fig.height=2}
DefaultAssay(pbmc.combined.sct)

umap_int <- DimPlot(pbmc.combined.sct, reduction = "umap")
cowplot::plot_grid(umap_merge, umap_int, ncol=2, labels = c("Merged", "Integrated"))

#colour by cluster
DimPlot(pbmc.combined.sct, reduction = "umap", group.by= c("integrated_snn_res.0.5", "integrated_snn_res.0.8")) 

```

Plot some canonical PBMC cluster markers e.g. S100A9, S100A8, LYZ, IGKC, GNLY, IGHM, NKG7, IGLC2

Which assay should we plot them on? (try both Integrated and SCT, is there a difference?)
First plot on the Integrated assay

```{r, fig.width=8}
DefaultAssay(pbmc.combined.sct)
FeaturePlot(pbmc.combined.sct, features = c("S100A9", "S100A8", "LYZ", "IGKC", "GNLY", "IGHM", "NKG7", "IGLC2"))

```

Set default assay to "SCT" then do the same plot
```{r, fig.width=8}
DefaultAssay(pbmc.combined.sct) <- "SCT"
FeaturePlot(pbmc.combined.sct, features = c("S100A9", "S100A8", "LYZ", "IGKC", "GNLY", "IGHM", "NKG7", "IGLC2")) #there's much less noise using SCT dataset. Why?
```

## 4. Identify markers

We want to identify markers that distinguish each cluster. 
Look up `FindMarkers` and `FindAllMarkers`, which one should we use here?

Why are we changing the DefaultAssay?
Do we need to renormalise RNA Assay?

Use the "LR" test

```{r}
DefaultAssay(pbmc.combined.sct) <- "RNA" #because we want an unscaled and unfiltered list of genes 
Idents(pbmc.combined.sct) <- pbmc.combined.sct$integrated_snn_res.0.5 #assign the active identifier of the object as the cluster label at the resolution you want 
pbmc.combined.sct[[]] #check the metadata by checking whether seurat clusters matches the idents we need

#Note to self: maybe you need to log-normalize the data before you perform marker detection, something like this:
#pbmc.combined.sct <- NormalizeData(pbmc.combined.sct, assay = "RNA", normalization.method = "LogNormalize", )

allmarkers <- FindAllMarkers(pbmc.combined.sct, latent.vars = "orig.ident", test.use = "LR") #wilcox won't work in this case to regress out orig.ident (which is the batch effect)?; so instead logistic regression is used 

```

Filter the top 5 markers for each cluster
use `dply::group_by` and `slice_max()`
How do you define "top"?

```{r}
allmarkers
pbmc.combined.sct.markers.top5 <- allmarkers %>% 
  group_by(cluster) %>% 
  slice_max(order_by = avg_log2FC, n = 5) #you should also filter out the low p-values as well when you do it by yourself
pbmc.combined.sct.markers.top5


```

plot Heatmap or Dotplot of markers

try functions `Seurat::DoHeatmap` or `Seurat::DotPlot`

```{r}

DefaultAssay(pbmc.combined.sct) <- "SCT" #need to do this on the SCT for some reason

Seurat::DoHeatmap(pbmc.combined.sct, features = pbmc.combined.sct.markers.top5$gene)

Seurat::DotPlot(pbmc.combined.sct, features = unique(pbmc.combined.sct.markers.top5$gene)) + theme(axis.text.x = element_text(angle=90))
```

How well do you think the clustering has worked?
Are you able to identify each cluster?
Would you re-evaluate any of the clusters?


## Optional 

Spend some time experimenting with different statistical tests in the FindAllMarkers command
Try some different parameters
Sort by log2fc vs padj
How does this affect the plots?
Use your results to label the identity of each cluster and plot on a UMAP

# Bonus Exercise: Run Harmony based integration

You can run Harmony within your Seurat workflow. You'll only need to make two changes to your code.

IMPORTANT DIFFERENCE: In the Seurat integration tutorial, you need to define a Seurat object for each dataset. 
With Harmony integration, create only one Seurat object with all cells.
Before running Harmony, make a Seurat object and following the standard pipeline through PCA.

After that Run Harmony with the `RunHarmony()` function
In downstream analyses, use the Harmony embeddings instead of PCA.
For example, run Harmony and then UMAP in two lines.

Vignette: http://htmlpreview.github.io/?https://github.com/immunogenomics/harmony/blob/master/docs/SeuratV3.html


Create a merged object, run SCTransform and PCA
```{r}
seurat_merged <- merge() #already did above for Seurat

```


Run Harmony and include the optimisation graph

```{r}
seurat_merged[[]]
seurat_merged_harmony <- RunHarmony(seurat_merged, reduction = "pca", group.by.vars = "orig.ident", dims.use = 1:20, plot_convergence = T)

```


Run UMAP, FindNeighbours and FindClusters, then visualise some key genes. 
```{r}
#Harmony gives you a separate reduction but not a separate assay, so you don't needa switch between assays

seurat_merged_harmony <- RunUMAP(seurat_merged_harmony, reduction = "harmony", dims = 1:20)
seurat_merged_harmony <- FindNeighbors(seurat_merged_harmony, reduction = "harmony", dims = 1:20)
seurat_merged_harmony <- FindClusters(seurat_merged_harmony)
```

How do the Harmony and Seurat integration methods compare?

```{r, fig.width=5, fig.height=2}


umap_harmony_int <- DimPlot(seurat_merged_harmony, reduction = "harmony", label = T)
seurat_merged_harmony[[]]
seurat_int_0.8 <- DimPlot(pbmc.combined.sct, reduction = "umap", group.by="integrated_snn_res.0.8", label = T) 

cowplot::plot_grid(seurat_int_0.8, umap_harmony_int, ncol=2, labels = c("Seurat", "Harmony"))
```


Compute LISI : https://github.com/immunogenomics/LISI 

